Orange/widgets/utils/listfilter.py

```diff
@@ -0,0 +1,244 @@
+    def __init__(self, variable_list_model=VariableListModel(enable_dnd=True)):
+        self.original_completer_items = []
+        super().__init__()
+        self.filter_edit = QLineEdit()
+        self.filter_edit.setToolTip("Filter the list of available variables.")
+
+        if hasattr(self.filter_edit, "setPlaceholderText"):
+            self.filter_edit.setPlaceholderText("Filter")
+
+        self.completer = QCompleter()
+        self.completer.setCompletionMode(QCompleter.InlineCompletion)
+        self.completer_model = QStringListModel()
+        self.completer.setModel(self.completer_model)
+        self.completer.setModelSorting(
+            QCompleter.CaseSensitivelySortedModel)
+
+        self.filter_edit.setCompleter(self.completer)
+        self.completer_navigator = CompleterNavigator(self)
+        self.filter_edit.installEventFilter(self.completer_navigator)
+
+        self.available_attrs = variable_list_model
+        self.available_attrs_proxy = VariableFilterProxyModel()
+        self.available_attrs_proxy.setSourceModel(self.available_attrs)
+        self.available_attrs_view = VariablesListItemView(
+            acceptedType=Variable)
+        self.available_attrs_view.setModel(self.available_attrs_proxy)
+
+        aa = self.available_attrs
+        aa.dataChanged.connect(self.update_completer_model)
+        aa.rowsInserted.connect(self.update_completer_model)
+        aa.rowsRemoved.connect(self.update_completer_model)
+
+        self.filter_edit.textChanged.connect(self.update_completer_prefix)
+        self.filter_edit.textChanged.connect(
+            self.available_attrs_proxy.set_filter_string)
+
+    def update_completer_model(self, *_):
+        """ This gets called when the model for available attributes changes
+        through either drag/drop or the left/right button actions.
+
+        """
+        vars = list(self.available_attrs)
+        items = [var.name for var in vars]
+        items += ["%s=%s" % item for v in vars for item in v.attributes.items()]
+
+        new = sorted(set(items))
+        if new != self.original_completer_items:
+            self.original_completer_items = new
+            self.completer_model.setStringList(self.original_completer_items)
+
+    def update_completer_prefix(self, filter):
+        """ Prefixes all items in the completer model with the current
+        already done completion to enable the completion of multiple keywords.
+        """
+        prefix = str(self.completer.completionPrefix())
+        if not prefix.endswith(" ") and " " in prefix:
+            prefix, _ = prefix.rsplit(" ", 1)
+            items = [prefix + " " + item
+                     for item in self.original_completer_items]
+        else:
+            items = self.original_completer_items
+        old = list(map(str, self.completer_model.stringList()))
+
+        if set(old) != set(items):
+            self.completer_model.setStringList(items)
+
+    def add_filter(self):
```

Comment:
Why is a method that *returns* a filter called `add_filter`? And a method that returns a model is called `add_model`? Qt calls such getters simply `filter`, `model`...

These methods won't be needed anyway if the class is changed into a function.